model/common/src/icon4py/model/common/grid/grid_refinement.py

```diff
@@ -0,0 +1,183 @@
+    dims.CellDim: (0, -4),
+    dims.EdgeDim: (0, -8),
+    dims.VertexDim: (0, -4),
+}
+"""Refinement value indicating a point is in the unordered interior (fully prognostic) of the grid: this is encoded by 0 or -4 in coarser parent grid."""
+
+_MIN_ORDERED: Final[dict[gtx.Dimension, int]] = {
+    dim: value[1] + 1 for dim, value in _UNORDERED.items()
+}
+"""For coarse parent grids the refinement control value of overlapping boundary regions are counted with negative values.
+"""
+
+
+DEFAULT_GRF_NUDGEZONE_WIDTH: Final[int] = 8
+
+
+_LAST_NUDGING: dict[gtx.Dimension, h_grid.Zone] = {
+    dims.EdgeDim: h_grid.Zone.NUDGING_LEVEL_2,
+    dims.CellDim: h_grid.Zone.NUDGING,
+    dims.VertexDim: h_grid.Zone.NUDGING,
+}
+_LAST_BOUNDARY: dict[gtx.Dimension, h_grid.Zone] = {
+    dims.EdgeDim: h_grid.Zone.LATERAL_BOUNDARY_LEVEL_8,
+    dims.CellDim: h_grid.Zone.LATERAL_BOUNDARY_LEVEL_4,
+    dims.VertexDim: h_grid.Zone.LATERAL_BOUNDARY_LEVEL_4,
+}
+
+
+def compute_domain_bounds(
+    dim: gtx.Dimension, refinement_fields: dict[gtx.Dimension, gtx.Field], array_ns: ModuleType = np
+) -> tuple[dict[h_grid.Domain, gtx.int32], dict[h_grid.Domain, gtx.int32]]:  # type: ignore   [name-defined]
+    refinement_ctrl = refinement_fields[dim].ndarray
+    refinement_ctrl = convert_to_unnested_refinement_values(refinement_ctrl, dim, array_ns)
+
+    domains = h_grid.get_domains_for_dim(dim)
+    start_indices = {}
+    end_indices = {}
+    for domain in domains:
+        start_index = 0
+        end_index = refinement_ctrl.shape[0]
+        my_zone = domain.zone
+        if (
+            my_zone is h_grid.Zone.END or my_zone.is_halo()
+        ):  # TODO(halungge): implement for distributed
+            start_index = refinement_ctrl.shape[0]
+            end_index = refinement_ctrl.shape[0]
+        elif my_zone.is_lateral_boundary():
+            found = array_ns.where(refinement_ctrl == my_zone.level)[0]
+            start_index, end_index = (
+                (array_ns.min(found).item(), array_ns.max(found).item() + 1)
+                if found.size > 0
+                else (0, 0)
+            )
+        elif my_zone.is_nudging():
+            value = _LAST_BOUNDARY[dim].level + my_zone.level
+            found = array_ns.where(refinement_ctrl == value)[0]
+            start_index, end_index = (
+                (array_ns.min(found).item(), array_ns.max(found).item() + 1)
+                if found.size > 0
+                else (0, 0)
+            )
+        elif my_zone is h_grid.Zone.INTERIOR:
+            # for the Vertex and Edges the level after the nudging zones are not ordered anymore, so
+            # we rely on using the end index of the nudging zone for INTERIOR
+            value = get_nudging_refinement_value(dim)
+            found = array_ns.where(refinement_ctrl == value)[0]
+            start_index = array_ns.max(found).item() + 1 if found.size > 0 else 0
+            end_index = refinement_ctrl.shape[0]
```

Comment:
What happens if we fall through? Then
```
        start_index = 0
        end_index = refinement_ctrl.shape[0]
```
is what we use, aka all-the-points, right?

Reply:
I.e. this algorithms is: we take the whole domain, then for same cases we shrink it?

Reply:
yes, kind of.  For a non limited area grid there is nothing other than "all points" except for halos, but the halos are not yet taken into account here. (See TODO above) The problem is that due to the ordering that they do or don't in the limited area case it is far less standard than I originall thought, or I did not get the right hook to implement it in a decent way.